cleanup spack.parse.Lexer

cosmicexplorer opened the pull request:
### Problem
The `SpecLexer` works fine, but is a little hard to extend, which we may wish to do when expanding the spec syntax as in #24025.

### Solution
- Rewrite the `Lexer` interface to allow navigation between multiple lexer modes by name.
- Also avoid the need to manually write out `lambda scanner, val: ...` in subclasses of `Lexer` for a more declarative interface.

### Result
The spec lexing code should be a little easier to follow. The path to extend the spec syntax as proposed in https://github.com/spack/spack/discussions/20256#discussioncomment-248931 should also be more clear now.

cosmicexplorer committed:
Make it possible to have more than one lexer mode at once!

cosmicexplorer committed:
Enable named instead of numeric lexer modes!

cosmicexplorer committed:
add types to Lexer

cosmicexplorer committed:
make parser operate on a token stream

cosmicexplorer committed:
remove extraneous self.token(...) calls in SpecLexer.__init__()

cosmicexplorer committed:
use re.escape() for most tokens instead of manual escapes

cosmicexplorer committed:
add back missing VER token and run black

cosmicexplorer committed:
break out confusing lines of code into their own variables and fix bug

cosmicexplorer committed:
fix py2 compat weirdness with type decl

cosmicexplorer committed:
add a lot more documentation of how we construct/execute the scanner

cosmicexplorer committed:
fix weird mypy error

cosmicexplorer commented on lib/spack/spack/spec.py:
```diff
@@ -4995,53 +4995,74 @@ def __missing__(self, key):
                 (
-                    r"\@([\w.\-]*\s*)*(\s*\=\s*\w[\w.\-]*)?",
-                    lambda scanner, val: self.token(VER, val),
+                    "BEGIN_PHASE",
```
This bit is really the entire point of this change. Essentially, instead of manually typing out lambdas in our actual lexer specification, we have a more declarative interface.

cosmicexplorer commented on lib/spack/spack/spec.py:
```diff
@@ -4995,53 +4995,74 @@ def __missing__(self, key):
+_windows_filename_prefix = r"([A-Za-z]:)*?"
+# Spec strings require posix-style paths on Windows
+# because the result is later passed to shlex.
+_windows_filename_re = _windows_filename_prefix + _filename_re_base
```
I haven't changed any of the actual strings here whatsoever, I've just moved them to the top level.

cosmicexplorer commented on lib/spack/spack/spec.py:
```diff
@@ -4995,53 +4995,74 @@ def __missing__(self, key):
+                    "BEGIN_PHASE",
+                    [
+                        # '@': begin a Version, VersionRange, or VersionList:
+                        (r"@([\w.\-]*\s*)*(\s*\=\s*\w[\w.\-]*)?", VER),
```
The only actual change I made to any of these strings is to remove the backslash from `r"\@"`  here since the backslash is simply ignored for any non-metacharacters.

cosmicexplorer commented on lib/spack/spack/spec.py:
```diff
@@ -4995,53 +4995,74 @@ def __missing__(self, key):
+                        # Gobble up all remaining whitespace between tokens.
+                        (r"\s+", None),
+                    ],
+                    {"EQUALS_PHASE": [EQ]},
```
This was the other important reason for this change. imo it was really unclear how the list of tuples we have now ends up triggering changes in the lexer mode, because we were relying on the implementation detail of only allowing exactly two modes.

While we do not add any new modes in this specific change, this makes it a little easier for us to extend the spec syntax later e.g. by having more complex compiler specs. I recall @alalazo was working on exactly that (how to add a new element to spec syntax for compiler specs) a while ago and I think the refactoring in this PR will make that effort easier in the future.

cosmicexplorer replied:
I was referring to this element of the virtual provider selection changes at https://github.com/spack/spack/pull/15569#issuecomment-1262947139!

cosmicexplorer commented:
`codecov/project` is failing, but `codecov/patch` is not. I could definitely add some further test cases triggering `LexError` if that would be helpful, but I was hoping that the existing tests would be enough to validate this refactoring change.

cosmicexplorer commented:
Every test is passing except `unit-tests / macos (3.8)`: https://github.com/spack/spack/actions/runs/3155071691/jobs/5133403157, which I expect is intermittent or not related to my code. Therefore, this is ready for review.

becker33 commented:
I really like the big picture here!

In the details, I think there's a better way to set up the constructor for the Lexer object. I don't have time to get into the weeds of this before I leave this evening but I'll try to take a look at the details soon.

cosmicexplorer commented:
@becker33: Yes! Would really love input on the Lexer ctor since the purpose of this change is to make that better. No rush to get back.

cosmicexplorer commented:
Noting a discussion with @trws and @psakievich over slack just now on how it's kinda difficult to handle flags reliably, stemming from https://github.com/spack/spack/pull/32257. In working on this change yesterday I was struck with how when you pass a string to the `Spec('%clang cxxflags="-stdlib=c++"')` constructor, it first shlexes it (which removes the double quotes), then sends each individual element to the lexer as `SpecLexer().lex(['%clang', 'cxxflags=-stdlib=c++'])`.

I absolutely think that the lexer should be doing the splitting here itself and I believe that doing that might make it easier to rely on specs being able to safely serialize and deserialize without breaking. I probably will not try to do it in this change, but in a separate one that modifies the control flow between the lexer and the parser.

cosmicexplorer commented:
Super helpful rundown from @becker33 on the lex/parse architectural constraints:

> So the shlex craziness comes from my original PR to support compiler flags — the easy way to do it required the argument to the lexer to arrive quoted, which required escaped quotes on the command line. That was doable without any shlex madness, but we wanted users to be able to use normal quotes on the command line, without losing the ability to round-trip specs through the str method without losing compiler flag information
>
> That required some way to unify the parsing between things that come in as a single string (internally from within python) and things that come in as CLI args
>
> shlex madness was my solution to that at the time (2015-16)

If there's any useful refactoring I can figure out on that front (in a separate PR) I'll keep this goal in mind.

cosmicexplorer commented:
@tgamblin: I believe the overall runtime regresses ~0.56%, and the performance of `lex_word` appears to be the culprit, probably due to the additional overhead of indexing into a dict when switching modes. I think we should still take this change, since the regression is extremely small and the resulting code seems more readable and maintainable. However, I am somewhat confident that removing the `shlex.split()` preprocessing (as per above comments) would make up for the regression, so I can look into that before we merge this change if you'd like.

*Methodology and perf regression specifics below.* **Important points in bold.**

### Overall Runtime

**Wall-clock time seems to stay mostly constant at around 23.5 seconds, regressing ~.1318 seconds, or 0.56%.** This PR's mean time was 23.4924, while `develop` was slightly lower at 23.3606 over 5 consecutive runs:
```bash
# cosmicexplorer@terrestrial-gamma-ray-flash: ~/tools/s1 23:55:33 
; for _ in $(seq 5); do time (spack python -c 'import spack.repo; list(spack.repo.path.all_package_classes())' >/dev/null); done
  22.75s user 0.38s system 100% cpu 23.133 total
( spack python -c  > /dev/null; )  22.75s user 0.38s system 99% cpu 23.133 total
  23.18s user 0.36s system 100% cpu 23.535 total
( spack python -c  > /dev/null; )  23.18s user 0.36s system 99% cpu 23.536 total
  23.19s user 0.38s system 99% cpu 23.571 total
( spack python -c  > /dev/null; )  23.19s user 0.38s system 99% cpu 23.572 total
  23.19s user 0.37s system 99% cpu 23.555 total
( spack python -c  > /dev/null; )  23.19s user 0.37s system 99% cpu 23.556 total
  23.30s user 0.37s system 99% cpu 23.665 total
( spack python -c  > /dev/null; )  23.30s user 0.37s system 99% cpu 23.665 total
# cosmicexplorer@terrestrial-gamma-ray-flash: ~/tools/s1 23:57:50 
; git checkout develop
Switched to branch 'develop'
Your branch is up to date with 'upstream/develop'.
# cosmicexplorer@terrestrial-gamma-ray-flash: ~/tools/s1 23:57:55 
; for _ in $(seq 5); do time (spack python -c 'import spack.repo; list(spack.repo.path.all_package_classes())' >/dev/null); done
  23.14s user 0.35s system 100% cpu 23.486 total
( spack python -c  > /dev/null; )  23.14s user 0.35s system 100% cpu 23.487 total
  23.14s user 0.35s system 100% cpu 23.488 total
( spack python -c  > /dev/null; )  23.14s user 0.35s system 99% cpu 23.489 total
  22.63s user 0.36s system 99% cpu 22.985 total
( spack python -c  > /dev/null; )  22.63s user 0.36s system 99% cpu 22.985 total
  22.91s user 0.37s system 99% cpu 23.282 total
( spack python -c  > /dev/null; )  22.91s user 0.37s system 99% cpu 23.283 total
  23.19s user 0.37s system 100% cpu 23.558 total
( spack python -c  > /dev/null; )  23.19s user 0.37s system 99% cpu 23.559 total
```

### Method-level Profiling
In this section we try to identify where that additional ~.13 seconds comes from, and can tentatively attribute it to `lex_word`.

**If this change is correct, none of the regex matching or `shlex` performance should change**, since the new `SpecLexer()` constructor should produce the exact same `re.Scanner` as on `develop`, and calls `shlex.split()` in the exact same way. If the performance *does* change, I would expect it to be in `parse.py` methods.

This hypothesis is tentatively proven correct below.

#### Profile Excerpt
I ran the same spack command as above with `-p --lines 40` to get the top 40 functions. I'm first going to show an excerpt of the outputs for just a few methods that seem important:
```
Branch: arbitrary-lexer-modes
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   152543    0.685    0.000    1.662    0.000 /usr/lib/python3.10/re.py:364(scan)
152543/115311    0.406    0.000    2.113    0.000 lib/spack/spack/parse.py:87(lex_word)
   440197    0.301    0.000    0.301    0.000 {method 'match' of '_sre.SRE_Scanner' objects}
   287654    0.269    0.000    0.405    0.000 lib/spack/spack/parse.py:76(token)
    99683    0.220    0.000    5.186    0.000 lib/spack/spack/parse.py:188(setup)

Branch: develop
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   152543    0.774    0.000    1.715    0.000 /usr/lib/python3.10/re.py:364(scan)
152543/115311    0.286    0.000    2.017    0.000 lib/spack/spack/parse.py:59(lex_word)
   440197    0.281    0.000    0.281    0.000 {method 'match' of '_sre.SRE_Scanner' objects}
   287654    0.263    0.000    0.394    0.000 lib/spack/spack/parse.py:53(token)
    99683    0.226    0.000    6.249    0.000 lib/spack/spack/parse.py:149(setup)
```

If I'm interpreting the `cProfile` output correctly, **`lex_word` goes from 0.286 to 0.406 seconds of total runtime (+0.12 seconds)** in the single profiled run I performed on both branches, **which aligns very closely with the regression in overall runtime (0.12 vs 0.1318)**. Especially since `lex_word` introduces a dict lookup in this PR vs an integer subtraction on `develop`, it makes sense that this method would produce the regression. Anecdotally, the other method runtimes seem to stay within the same distribution as I repeat profiled runs regardless of branch, while `lex_word` seems to reliably regress around 0.12 seconds each time.

#### Top 40 Profile Output

This is the precise command line, wall-clock time, and `cProfile` output I used to produce the above excerpt.

On `arbitrary-lexer-modes`:
```
; spack -p --lines 40 python -c 'import spack.repo; list(spack.repo.path.all_package_classes())'
...
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   214994    1.760    0.000    2.062    0.000 /usr/lib/python3.10/shlex.py:133(read_token)
    55885    1.394    0.000    2.987    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:3895(_dup_deps)
147481/145199    1.253    0.000   13.106    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:1188(__init__)
    99683    0.897    0.000    0.940    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/parse.py:136(__init__)
   152543    0.685    0.000    1.662    0.000 /usr/lib/python3.10/re.py:364(scan)
1932561/1774121    0.609    0.000    2.760    0.000 /home/cosmicexplorer/tools/s1/lib/spack/llnl/util/lang.py:301(<genexpr>)
543512/413500    0.589    0.000    3.238    0.000 /home/cosmicexplorer/tools/s1/lib/spack/llnl/util/lang.py:299(tuplify)
539311/539010    0.528    0.000    1.063    0.000 {built-in method builtins.sorted}
55896/55885    0.502    0.000    4.836    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:3803(_dup)
   203862    0.490    0.000    0.526    0.000 /usr/lib/python3.10/_collections_abc.py:839(values)
383198/164530    0.425    0.000    0.739    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/directives.py:277(remove_directives)
  2863848    0.412    0.000    0.475    0.000 {built-in method builtins.isinstance}
152543/115311    0.406    0.000    2.113    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/parse.py:87(lex_word)
67945/66791    0.379    0.000    1.703    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:1630(traverse_edges)
946895/946893    0.377    0.000    0.377    0.000 {built-in method builtins.iter}
   406754    0.374    0.000    0.374    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:832(__init__)
  1483350    0.352    0.000    0.538    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:4066(_cmp_iter)
    49933    0.347    0.000    0.398    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/version.py:1137(__eq__)
    58286    0.339    0.000   20.369    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/directives.py:357(_depends_on)
    99683    0.316    0.000    0.325    0.000 /usr/lib/python3.10/shlex.py:21(__init__)
  1086673    0.313    0.000    0.317    0.000 {built-in method builtins.setattr}
   440197    0.301    0.000    0.301    0.000 {method 'match' of '_sre.SRE_Scanner' objects}
    99683    0.294    0.000    1.234    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:5078(__init__)
   741440    0.282    0.000    0.934    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/parse.py:158(accept)
   101965    0.275    0.000    3.334    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:5268(spec)
   287654    0.269    0.000    0.405    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/parse.py:76(token)
   298198    0.268    0.000    1.070    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/version.py:939(__init__)
730956/150599    0.248    0.000    1.910    0.000 {built-in method builtins.hash}
    99683    0.244    0.000    4.079    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:5090(do_parse)
1508634/1037026    0.244    0.000    0.338    0.000 {built-in method builtins.len}
   192351    0.243    0.000    0.479    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/version.py:1171(_string_to_version)
    93991    0.228    0.000    0.489    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/version.py:254(_generate_seperators_and_components)
    99683    0.222    0.000    2.874    0.000 /usr/lib/python3.10/shlex.py:305(split)
    99683    0.220    0.000    5.186    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/parse.py:188(setup)
    82265    0.219    0.000    1.711    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/directives.py:248(_wrapper)
   201583    0.213    0.000    0.213    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/directives.py:285(<genexpr>)
270558/128316    0.205    0.000    0.776    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/version.py:107(coercing_method)
   530638    0.180    0.000    0.335    0.000 /usr/lib/python3.10/_collections_abc.py:904(__iter__)
  1018780    0.174    0.000    0.174    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/hash_types.py:31(attr)
   437395    0.174    0.000    0.243    0.000 /usr/lib/python3.10/_collections_abc.py:835(items)


  32.92s user 0.40s system 99% cpu 33.334 total
```

On `develop`:
```
; spack -p --lines 40 python -c 'import spack.repo; list(spack.repo.path.all_package_classes())'
...
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   214994    1.751    0.000    2.049    0.000 /usr/lib/python3.10/shlex.py:133(read_token)
147481/145199    1.189    0.000   12.720    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:1188(__init__)
55896/55885    1.108    0.000    4.482    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:3803(_dup)
    99683    1.080    0.000    1.089    0.000 /usr/lib/python3.10/shlex.py:21(__init__)
   152543    0.774    0.000    1.715    0.000 /usr/lib/python3.10/re.py:364(scan)
   203377    0.727    0.000    0.759    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:745(__init__)
1932561/1774121    0.587    0.000    2.653    0.000 /home/cosmicexplorer/tools/s1/lib/spack/llnl/util/lang.py:301(<genexpr>)
543512/413500    0.572    0.000    3.117    0.000 /home/cosmicexplorer/tools/s1/lib/spack/llnl/util/lang.py:299(tuplify)
383198/164530    0.565    0.000    0.866    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/directives.py:277(remove_directives)
270558/128316    0.557    0.000    0.803    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/version.py:107(coercing_method)
    55885    0.473    0.000    1.607    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:3895(_dup_deps)
  2863848    0.399    0.000    0.462    0.000 {built-in method builtins.isinstance}
   192351    0.391    0.000    0.625    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/version.py:1171(_string_to_version)
539311/539010    0.382    0.000    0.890    0.000 {built-in method builtins.sorted}
67945/66791    0.377    0.000    1.238    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:1630(traverse_edges)
    69008    0.344    0.000    0.525    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:880(select)
  1483350    0.336    0.000    0.518    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:4066(_cmp_iter)
    82265    0.333    0.000    1.930    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/directives.py:248(_wrapper)
    58286    0.323    0.000   20.141    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/directives.py:357(_depends_on)
  1086673    0.302    0.000    0.306    0.000 {built-in method builtins.setattr}
152543/115311    0.286    0.000    2.017    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/parse.py:59(lex_word)
   440197    0.281    0.000    0.281    0.000 {method 'match' of '_sre.SRE_Scanner' objects}
   741440    0.275    0.000    0.459    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/parse.py:119(accept)
   101965    0.268    0.000    3.069    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:5247(spec)
   287654    0.263    0.000    0.394    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/parse.py:53(token)
   298198    0.262    0.000    1.196    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/version.py:939(__init__)
730956/150599    0.241    0.000    1.825    0.000 {built-in method builtins.hash}
    99683    0.235    0.000    3.728    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:5069(do_parse)
    93991    0.230    0.000    0.485    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/version.py:254(_generate_seperators_and_components)
1508634/1037026    0.229    0.000    0.319    0.000 {built-in method builtins.len}
    99683    0.226    0.000    6.249    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/parse.py:149(setup)
    99683    0.217    0.000    3.618    0.000 /usr/lib/python3.10/shlex.py:305(split)
   406754    0.213    0.000    0.213    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:832(__init__)
   201583    0.203    0.000    0.203    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/directives.py:285(<genexpr>)
    66627    0.186    0.000    0.723    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/spec.py:1343(edges_to_dependencies)
1046578/1046576    0.179    0.000    0.179    0.000 {built-in method builtins.iter}
   530638    0.172    0.000    0.361    0.000 /usr/lib/python3.10/_collections_abc.py:904(__iter__)
  1018780    0.171    0.000    0.171    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/hash_types.py:31(attr)
   203377    0.167    0.000    0.205    0.000 /home/cosmicexplorer/tools/s1/lib/spack/spack/variant.py:534(__init__)
   437395    0.165    0.000    0.227    0.000 /usr/lib/python3.10/_collections_abc.py:835(items)


  31.87s user 0.37s system 99% cpu 32.247 total
```

cosmicexplorer commented:
I think we can pause reviewing this one for now, since the biggest benefit of this PR would be to enable #33016. I still have a bit more to do on that one, but when I can demonstrate that #33016 conclusively improves spec parsing, I think it will be easier to sign off on this component of that larger change.

alalazo commented:
Superseded by #34151
